### PR TITLE
Feature/minify

### LINKED ISSUE #24: No longer deploy source
You can use gulp, grunt or even just npm to run build tasks that can consolidate js files, reducing the number of round trips that pages have to make. You can also incorporate transpiling to earlier JS version, expanding the supported devices

### PR DESCRIPTION
**Pull request by GustavoFernandes**
Fixes #24

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -3,40 +3,86 @@ var deployDir = 'docs';
...
+});
+
+gulp.task('html', function () {
+  var target = gulp.src('src/index.html');
+  var sources = gulp.src(['src/**/*.js', 'src/**/*.css']);
+
+  return target.pipe(inject(sources, {
+    relative:true
+  }))
+      .pipe(gulp.dest('src'));
```
Comment (jonsmithers) [inline gulpfile.js]: `gulp html` modifies `src/index.html`? It injects all bower_components 'js' and 'css' files regardless of whether or not we explicitly import them?

File: gulpfile.js [hunk 2]
```diff
@@ -3,40 +3,86 @@ var deployDir = 'docs';
...
 });
 
-gulp.task('clean', function() {
-    return gulp.src(deployDir)
-        .pipe(clean());
+gulp.task('build-js', ['clean'], function () {
+  return gulp.src(src.map(function (path) {
+    return path + '/*.js';
```
Comment (jonsmithers) [inline gulpfile.js]: `src.map(path => path + '/*.js')`

File: gulpfile.js (hunk 1 above)
Comment (GustavoFernandes) [inline gulpfile.js]: Removed `gulp html` task as it served no real purpose.

File: gulpfile.js (hunk 2 above)
Comment (GustavoFernandes) [inline gulpfile.js]: Switched to using those arrow functions instead

File: src/index.html [hunk 3]
```diff
@@ -2,11 +2,14 @@
...
+		<!-- inject:js -->
 		<script src="main.js"></script>
-		<script src="parsers.js"></script>
 		<script src="order.js"></script>
-		<script src="install.js"></script>
+		<script src="parsers.js"></script>
+		<!-- endinject -->
+		<!-- inject:css -->
 		<link rel="stylesheet" href="styles.css">
+		<!-- endinject -->
```
Comment (GustavoFernandes) [inline src/index.html]: This render-blocking issue was not introduced as a result of this pull request so I'd recommend opening a new issue to address this.

File: gulpfile.js [hunk 4]
```diff
@@ -3,40 +3,75 @@ var deployDir = 'docs';
...
+var uglify = require('gulp-uglify');
+var babel = require('gulp-babel');
+var inject = require('gulp-inject');
+var minifyHtml = require('gulp-minify-html');
+var minifyCss = require('gulp-clean-css');
+var browserSync = require('browser-sync');
+var injectVersion = require('gulp-inject-version');
 
-gulp.task('default', ['copy']);
+gulp.task('default', ['lint', 'html']);
```
Comment (jonsmithers) [inline gulpfile.js]: The html task no longer exists? So now the default gulp task just lint's your code.

File: gulpfile.js [hunk 5]
```diff
@@ -3,40 +3,75 @@ var deployDir = 'docs';
...
+      }))
+      .pipe(jshint.reporter('jshint-stylish'))
+      .pipe(jshint.reporter('fail'));
 });
 
-gulp.task('clean', function() {
-    return gulp.src(deployDir)
-        .pipe(clean());
+gulp.task('build-js', ['clean'], function () {
+  return gulp.src(src.map(path => path + '/*.js'))
```
Comment (jonsmithers) [inline gulpfile.js]: Shouldn't the `sw/` directory be _inside_ `src/`? Also, I think `sw.js` is supposed injected into `index.html`. Only `sw/install.js` should be injected.

File: gulpfile.js [hunk 6]
```diff
@@ -3,40 +3,75 @@ var deployDir = 'docs';
...
-        .pipe(jshint.reporter('fail'));
+gulp.task('build-css', ['clean'], function () {
+  return gulp.src('src/**/*.css')
+      .pipe(minifyCss())
+      .pipe(gulp.dest(deployDir));
+});
+
+gulp.task('build-html', ['build-js', 'build-css'], function () {
+  var target = gulp.src('src/index.html');
+  var sources = gulp.src(deployDir + '/*.{js,css}');
```
Comment (GustavoFernandes) [inline gulpfile.js]: `gulp-inject` only injects the file references (script and link tags) not the entire file. All the files in the `docs/` directory are important.

The JS and CSS files in `docs/` don't get modified at this point; it only gets a list of the files paths (I'll add `{read: false}` to the `gulp.src` to enforce that).

And I would think that if/when we bring in third-party libraries and components, we can update the gulpfile to make it easier to work with.

File: gulpfile.js (hunk 5 above)
Comment (GustavoFernandes) [inline gulpfile.js]: The reason `sw/` is outside `src/` is so it doesn't get loaded into browsersync with `gulp serve`.

And every JS file gets concatenated and minified into `all.min.js` which later is injected as a reference into `index.html`.

File: gulpfile.js (hunk 4 above)
Comment (GustavoFernandes) [inline gulpfile.js]: Oops, I'll remove that. I think for now it's fine that the default gulp task just lints until something better gets presented.

File: gulpfile.js (hunk 5 above)
Comment (jonsmithers) [inline gulpfile.js]: I understand all.min.js, but sw.js doesn't belong there.

Comment (GustavoFernandes) [inline gulpfile.js]: Are you saying `sw.js` doesn't belong **in** `all.min.js`?

Comment (jonsmithers) [inline gulpfile.js]: Yes

File: gulpfile.js (hunk 6 above)
Comment (jonsmithers) [inline gulpfile.js]: Oh........so, like, all gulp-inject does is insert import tags for us? Why do we need that? Why not write our own import tags?

Comment (jonsmithers) [inline gulpfile.js]: I think I get it. It removes imports for unminified files and replaces them with imports for minified files.

Comment (jonsmithers) [inline gulpfile.js]: Yeah, OK nvm :-)

File: gulpfile.js (hunk 5 above)
Comment (GustavoFernandes) [inline gulpfile.js]: I don't understand why `sw.js` should not make it into `all.min.js`?

Comment (GustavoFernandes) [inline gulpfile.js]: I removed anything under `sw/` from the build and merged this pull request. I opened another issue to resolve this #42

**Pull request merged**